core/src/main/java/org/mule/util/MapUtils.java

```diff
@@ -138,4 +140,27 @@ public static String toString(Map props, boolean newline)
         buf.append('}');
         return buf.toString();
     }
+
+    /**
+     * Puts the {@code key}/{@code value} pair into the given {@code map} only as long
+     * as the {@code key} is not already present on the {@code map}. If the {@code key}
+     * is already present then a {@link IllegalStateException} is thrown.
+     *
+     * @param map   a {@link Map}
+     * @param key   the key
+     * @param value the value
+     * @param <K>   the generic type of the key
+     * @param <V>   the generic type of the value
+     * @throws IllegalStateException if the {@code map} already contains the {@code key}
+     */
+    public static <K, V> void idempotentPut(Map<K, V> map, K key, V value)
+    {
+        checkArgument(value != null, "value cannot be null");
+        if (map.containsKey(key))
+        {
+            throw new IllegalStateException(String.format("A %s is already registered for the name '%s'", value.getClass().getSimpleName(), key));
```

Comment:
A %s is already registered for the name '%s' > Key '%s' is already registered with value %s